sdmetrics/reports/single_table/base_report.py

```diff
@@ -0,0 +1,221 @@
+"""Single table base report."""
+import pickle
+import sys
+import warnings
+
+import numpy as np
+import pandas as pd
+import pkg_resources
+import tqdm
+
+from sdmetrics.reports.utils import _validate_categorical_values
+
+
+class BaseReport():
+    """Base report class for single table reports.
+
+    This class creates a base report for single-table data.
+    """
+
+    def __init__(self):
+        self._overall_score = None
+        self.is_generated = False
+        self._properties = {}
+
+    def _validate_metadata_matches_data(self, real_data, synthetic_data, metadata):
+        """Validate that the metadata matches the data.
+
+        Raise an error if the column metadata does not match the column data.
+
+        Args:
+            real_data (pandas.DataFrame):
+                The real data.
+            synthetic_data (pandas.DataFrame):
+                The synthetic data.
+            metadata (dict):
+                The metadata of the table.
+        """
+        real_columns = set(real_data.columns)
+        synthetic_columns = set(synthetic_data.columns)
+        metadata_columns = set(metadata['columns'].keys())
+
+        missing_data = metadata_columns.difference(real_columns.union(synthetic_columns))
+        missing_metadata = real_columns.union(synthetic_columns).difference(metadata_columns)
+        missing_columns = missing_data.union(missing_metadata)
+
+        if missing_columns:
+            error_message = (
+                'The metadata does not match the data. The following columns are missing'
+                ' in the real/synthetic data or in the metadata: '
+                f"{', '.join(sorted(missing_columns))}"
+            )
+            raise ValueError(error_message)
+
+    def validate(self, real_data, synthetic_data, metadata):
+        """Validate the inputs.
+
+        Args:
+            real_data (pandas.DataFrame):
+                The real data.
+            synthetic_data (pandas.DataFrame):
+                The synthetic data.
+            metadata (dict):
+                The metadata of the table.
+        """
+        if not isinstance(metadata, dict):
+            metadata = metadata.to_dict()
+
+        self._validate_metadata_matches_data(real_data, synthetic_data, metadata)
+        _validate_categorical_values(real_data, synthetic_data, metadata)
+
+    def _print_results(self, out=sys.stdout):
+        raise NotImplementedError
+
+    def _get_num_iterations(self, property_name, metadata):
+        raise NotImplementedError
+
+    def generate(self, real_data, synthetic_data, metadata, verbose=True):
+        """Generate report.
+
```

Comment:
Can we add a short description, eg: Iterate over all the `properties` and compute the scores etc..

Reply:
Yes, done in [fcae5dc](https://github.com/sdv-dev/SDMetrics/pull/411/commits/fcae5dc0a63db277f9b06746c587f2fdd0fb2cd0).